Feature esxi vcenter

toros11 opened the pull request:
This PR adds following features to the esxi driver

- Ability to use vCenter for instance creating
- Set the datastore location/base image name in json template (vmkfs cloning)
- Use ovf templates as the base for which to create new instances (vCenter)
- Use inbuilt clone feature (vCenter)
- Add nic/set network id
- Add floppy device used for metadata image if not present

b0r6 committed:
Added option for which clone method to use

b0r6 committed:
Use different clone method depending on config

b0r6 committed:
Added config option

b0r6 committed:
added image location to template

toros11 committed:
Set datastore and base image as separate parameters

Cloning from vCenter templates doesn't require any datastore as
it can look up the correct template by id. I also separated the
parts related to the machine into a separate file and cleaned up
some redundant variables/types

toros11 committed:
Let the executor configuration file handle vCenter flag

toros11 committed:
Restructure configuration/parameter handling

toros11 committed:
Refactor for vCenter/ESXi

vCenter can handle namin/registring of vm's through the clone
functionallity and does not need to go through these steps.

toros11 committed:
Added missing setting parameter

toros11 committed:
Parse the inner template message

toros11 committed:
Temporary disable metadrive related code

toros11 committed:
Update vendors

toros11 committed:
Fix mistakes in vendor update

toros11 committed:
Added field: id to esxi nic interface

toros11 committed:
Remove redundant function

toros11 committed:
Change nic types to match esxi/vcenter terminology

toros11 committed:
Store the nic as metadata regardless of type

toros11 committed:
Setup netowrk devices

toros11 committed:
Re-enable metadata code

toros11 committed:
Merge branch 'master' into feature-esxi-vcenter

toros11 committed:
Readded accedentally removed field and fix syntax

toros11 committed:
Add/Update esxi template

Created sample template for use with vCenter crated OVF templates

toros11 committed:
Added missing vendor

toros11 committed:
Fix another missing dependency

toros11 committed:
Update test

toros11 committed:
Update model

toros11 committed:
Use get functions to avoid nil pointer

toros11 committed:
Rename id -> network id

toros11 committed:
Update test case to include network id/esxi adapter type

toros11 committed:
Skip if the machine already has a nic inside the network

toros11 committed:
Fix typo

toros11 committed:
Update model

toros11 committed:
Forgot to update schema

toros11 committed:
Added missing character

toros11 committed:
Update vendor file

toros11 committed:
Wrote function to capture stdout

Vmwares library prints the output of api calls directly to stdout
and only returns an error code of wether the requests was able to
be sent succesfully or not. This lets us use the actual output of
the request to decide whether the request was succesfull or not

toros11 committed:
Parse device info to json format

toros11 committed:
Added missing dependency

toros11 committed:
Refactored the device check into a function

toros11 committed:
Wrote a function that checks if floppy device exists before adding

Currently openvdc uses fd0 to as mount point for metadata, here
we make sure that it gets attached correctly, and skips adding
a new device if one already exists

toros11 committed:
go fmt

toros11 committed:
Set esxi base image parameters directly in image structure

toros11 committed:
Fix some tests that i forgot to update

toros11 committed:
Fix some mistakes in template files

toros11 committed:
Create test that uses a base image without any devices attached

Also did some refactored the parameters into separate variables to
make it more readable

toros11 committed:
Allow image attributes to be overwritten

toros11 committed:
Added error type for govc api requests

Simply return when device.info fails.

b0r6 commented on pkg/conf/executor.toml:
```diff
@@ -13,8 +13,11 @@
 # esxi-datacenter = "ha-datacenter"
 # esxi-insecure = true
 # esxi-host-sshkey = ""
+# esxi-easy-clone = false
```
It seems like the options ```'esxi-easy-clone'``` and ```'esxi-basevm-datastore'``` isn't needed anymore.

toros11 committed:
Merge branch 'master' into feature-esxi-vcenter

toros11 committed:
Merge branch 'master' into feature-esxi-vcenter

toros11 committed:
Update sample config

unakatsuo commented on hypervisor/esxi/esxi.go:
```diff
@@ -97,6 +86,7 @@ func (p *EsxiHypervisorProvider) Name() string {
 func init() {
 	hypervisor.RegisterProvider("esxi", &EsxiHypervisorProvider{})
 	viper.SetDefault("hypervisor.esxi-insecure", true)
+	viper.SetDefault("hypervisor.vCenter", false)
```
This parameter breaks the existing naming convention. It would be better to start with ``esxi-`` prefix.

toros11 replied:
vCenter doesn't have anything to do with esxi though? It's just the endpoint.

Maybe it makes more sense to make a separate optional config field in the .toml file?

```
[vcenter]
enabled = true/false
```

unakatsuo commented on hypervisor/esxi/esxi.go:
```diff
@@ -216,6 +247,7 @@ func runCmd(cmd string, args []string) error {
 	return nil
 }
 
+var ErrApiRequest = errors.New("Failed api requiest")
```
typo...

toros11 committed:
Merge branch 'master' into feature-esxi-vcenter

unakatsuo commented on hypervisor/esxi/esxi.go:
```diff
@@ -208,6 +204,41 @@ func join(separator byte, args ...string) string {
 }
 
+func captureStdout(fn func() error) ([]byte, error) {
+	r, w, err := os.Pipe()
```
``r`` and ``w`` should be closed properly until it returns from this function.

toros11 committed:
Close/restore file descriptors before function return

toros11 committed:
fix typo

toros11 committed:
Don't use defer to close writer

The writer needs to close explicitly in order for io.Copy() to
return something, the writer needs to stay open until the buffer
is returned

toros11 committed:
Added unit test for captureStdout() function

toros11 committed:
Added modules subfield to config files

Using vCenter as the endpoint means that there is possibilities of
integration with a datacenter. This should make the management of
future features easier

toros11 committed:
Merge branch 'master' into feature-esxi-vcenter

toros11 committed:
Revert and add prefix to vcenter flag

Metallion commented:
Merging since @b0r6 has approved, @unakatsuo's not available for review and there is a commit addressing his comment.
